src/algorithm/REDUCE_SUM-Hip.cpp

```diff
@@ -35,37 +35,23 @@ namespace algorithm
 #define REDUCE_SUM_DATA_TEARDOWN_HIP \
   deallocHipDeviceData(x);
 
+#define REDUCE_SUM_BODY_HIP(atomicAdd) \
+  RAJAPERF_REDUCE_1_HIP(Real_type, REDUCE_SUM_VAL, dsum, sum_init, RAJAPERF_ADD_OP, atomicAdd)
+
 template < size_t block_size >
 __launch_bounds__(block_size)
 __global__ void reduce_sum(Real_ptr x, Real_ptr dsum, Real_type sum_init,
                            Index_type iend)
 {
-  HIP_DYNAMIC_SHARED(Real_type, psum);
-
-  Index_type i = blockIdx.x * block_size + threadIdx.x;
-
-  psum[ threadIdx.x ] = sum_init;
-  for ( ; i < iend ; i += gridDim.x * block_size ) {
-    psum[ threadIdx.x ] += x[i];
-  }
-  __syncthreads();
-
-  for ( i = block_size / 2; i > 0; i /= 2 ) {
-    if ( threadIdx.x < i ) {
-      psum[ threadIdx.x ] += psum[ threadIdx.x + i ];
-    }
-     __syncthreads();
-  }
+  REDUCE_SUM_BODY_HIP(::atomicAdd)
+}
 
-#if 1 // serialized access to shared data;
-  if ( threadIdx.x == 0 ) {
-    RAJA::atomicAdd<RAJA::hip_atomic>( dsum, psum[ 0 ] );
-  }
-#else // this doesn't work due to data races
-  if ( threadIdx.x == 0 ) {
-    *dsum += psum[ 0 ];
-  }
-#endif
+template < size_t block_size >
+__launch_bounds__(block_size)
+__global__ void reduce_sum_unsafe(Real_ptr x, Real_ptr dsum, Real_type sum_init,
```

Comment:
If we start to add more tests with atomics we might not want to splitting these up into separate kernel for compile time concerns but should be OK for now